### PR TITLE
Refactor and update the data pipeline

### PR DESCRIPTION
**Pull request by Skydes**
- Improved modularity in the model and dataset modules;
- Revised data augmentation process;
- Support for homographies with bordering artifacts;
- Results on SyntheticShapes.

See the individual commit comments for more details.

### CODE REVIEW

File: superpoint/models/utils.py [hunk 1]
```diff
@@ -64,6 +47,67 @@ def descriptor_head(inputs, **config):
...
+
+    # Mask the pixels if bordering artifacts appear
+    valid_mask = tf.ones_like(keypoint_map) if valid_mask is None else valid_mask
+    valid_mask = tf.to_float(valid_mask[..., tf.newaxis])  # for GPU
+    valid_mask = tf.space_to_depth(valid_mask, config['grid_size'])
+    valid_mask = tf.reduce_prod(valid_mask, axis=3)  # AND along the channel dim
+
+    loss = tf.losses.sparse_softmax_cross_entropy(labels=labels, logits=logits)
+    loss = tf.reduce_sum(valid_mask * loss) / tf.reduce_sum(valid_mask)
+    loss /= tf.to_float(tf.shape(logits)[-1])
```
Comment (rpautrat) [inline superpoint/models/utils.py]: Why do you need to divide by this?
The previous line `loss = tf.reduce_sum(valid_mask * loss) / tf.reduce_sum(valid_mask)` should already have computed the right mean, shouldn't it?

File: superpoint/models/utils.py [hunk 2]
```diff
@@ -105,7 +106,18 @@ def descriptor_loss(descriptors, warped_descriptors, homographies, **config):
...
+    # Mask the pixels if bordering artifacts appear
+    valid_mask = tf.ones([batch_size, Hc, Wc], tf.float32)\
+        if valid_mask is None else valid_mask
+    valid_mask = tf.to_float(valid_mask[..., tf.newaxis])  # for GPU
+    valid_mask = tf.space_to_depth(valid_mask, config['grid_size'])
+    valid_mask = tf.reduce_prod(valid_mask, axis=3)  # AND along the channel dim
+    valid_mask = tf.reshape(valid_mask, [batch_size, 1, 1, Hc, Wc])
+
+    normalization = tf.reduce_sum(valid_mask) * tf.reduce_sum(valid_mask) / batch_size
+    loss = tf.reduce_sum(valid_mask * loss) / normalization
```
Comment (Skydes) [inline superpoint/models/utils.py]: When broadcasting, the shape of `valid_mask` is changed from `[B, 1, 1, Hc, Wc]` to `[B, Hc, Wc, Hc, Wc]`. Thus shouldn't the normalization factor be `tf.reduce_sum(valid_mask) * Hc * Wc` instead ?

Comment (rpautrat) [reply superpoint/models/utils.py]: Indeed, good point.
